### DISCUSSION TITLE
Link slash causes a 404

### DISCUSSION THREAD

**Discussion by camilian** [zitadel/zitadel#8190] [Q&A]
The service I am using adds a "/" at the end of the service link.  Which causes a 404 error in 
zaitadel. 

`https://******.zitadel.cloud/oauth/v2/authorize/?client_id=***`

Gives a 404, where 

`https://******.zitadel.cloud/oauth/v2/authorize?client_id=***`

works.  I have no control over the link to the service calls.  Is a way get Zitadel to accept the slash before the GET varibles?

**Comment by fforootd** [comment]
Hm what service are we taking here?

**Reply by camilian** [reply]
cribl.cloud is the service. For the settings, they have "Issuer URL" input but no other options on how it crafts the urls.

**Reply by fforootd** [reply]
Can you share with me what you configured and how the url looks like when you see a 404?

You can send an email to florian@zitadel.com

**Reply by camilian** [reply]
Hello.  The issue is caused by the slash before the GET variables in the links the service uses. 
If the slash is removed the zitadel site will accept it. 

If I take a link that is created by the service for testing - 

`https://******.zitadel.cloud/oauth/v2/authorize/?client_id=***`

Remove the /  before the GET vars - 
`
https://******.zitadel.cloud/oauth/v2/authorize?client_id=***`

zitadel.cloud no longer responds with the 404.  I have opened up a ticket with the service also.  While I think the client shouldn't add the link, it would make sense for zitadel to work with both /authorize/?client_id=  **and** /authorize?client_id=

**Reply by fforootd** [reply]
Hey thanks again for the detailed description 👍 

I wanted to check if there could be a possible problem with a trailing `/` from the issuer url.

@livio-a @muhlemmer what you think about this. I quickly checked the oauth/oidc specs. and I saw no clear rule about the trailing `/` handling.

**Reply by muhlemmer** [reply]
This is a particular thing with Go routers.  Most routers I know in the Go ecosystem treat a trailing slash as a seperate route. In some cases even as a prefix route to sub-routes.

The way that I read the URI [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986#section-3.3) each slash in the path starts a new segment.

- `/path` has one segment, where the resource is called `path`
- `/path/` has two segments, first one is called `path` and the second is on root of `path`.

In OIDC we use the `chi` router package which also explicitly states (in the package documentation [overview](https://pkg.go.dev/github.com/go-chi/chi/v5#pkg-overview)):

> Trailing slashes on paths must be handled explicitly.

At the moment we only allow configuration of one path per endpoint. Default is without trailing slash. If you really need trailing slashes, you can edit the endpoint configuration in zitadel's configuration as set here in `defaults.yaml`:

https://github.com/zitadel/zitadel/blob/cd6b452fc9f185142085eb9a9cd62197b0d27cb5/cmd/defaults.yaml#L329-L345

But in my opinion the client should obtain the endpoints using [OIDC Discovery](https://zitadel.com/docs/apis/openidoauth/endpoints#openid-connect-10-discovery) and not modify them as they see fit. It can lead to all sort of problems. For example in OIDC terms `https://issuer-domain.com` and `https://issuer-domain.com/` are not the same issuer.

**Reply by muhlemmer** [reply]
Aah, I noticed now this is zitadel cloud, so the endpoint config can't be changed for your particular instance. It's a global setting.
